GG/src/Font.cpp

```diff
@@ -548,6 +695,191 @@ StrSize Font::TextElement::StringSize() const
+        size_t tag_begin = m_text.size();
+        size_t tag_name_begin = m_text.append("</").size();
+        size_t tag_name_end = m_text.append(tag).size();
+        size_t tag_end = m_text.append(">").size();
+        element->text = Substring(m_text,
+                                  boost::next(m_text.begin(), tag_begin),
+                                  boost::next(m_text.begin(), tag_end));
+
+        element->tag_name = Substring(m_text,
+                                      boost::next(m_text.begin(), tag_name_begin),
+                                      boost::next(m_text.begin(), tag_name_end));
+        m_text_elements.push_back(element);
+    }
+
+    /** Add a text element.  Any whitespace in this text element will be non-breaking.*/
+    void AddText(const std::string& text)
+    {
+        m_are_widths_calculated = false;
+
+        boost::shared_ptr<Font::TextElement> element(new Font::TextElement(false, false));
+        size_t begin = m_text.size();
+        size_t end = m_text.append(text).size();
+        element->text = Substring(m_text,
+                                  boost::next(m_text.begin(), begin),
+                                  boost::next(m_text.begin(), end));
+        m_text_elements.push_back(element);
+    }
+
+    /** Add a white space element.*/
+    void AddWhitespace(const std::string& whitespace)
+    {
+        m_are_widths_calculated = false;
+
+        boost::shared_ptr<Font::TextElement> element(new Font::TextElement(true, false));
+        size_t begin = m_text.size();
+        size_t end = m_text.append(whitespace).size();
+        element->text = Substring(m_text,
+                                  boost::next(m_text.begin(), begin),
+                                  boost::next(m_text.begin(), end));
+        m_text_elements.push_back(element);
+    }
+
+    /** Add a white space element.*/
+    void AddNewline()
+    {
+        m_are_widths_calculated = false;
+
+        boost::shared_ptr<Font::TextElement> element(new Font::TextElement(false, true));
+        m_text_elements.push_back(element);
+    }
+
+    /** Add open color tag.*/
+    void AddOpenTag(const Clr& color) {
+        std::vector<std::string> params;
+        params.push_back(boost::lexical_cast<std::string>(static_cast<int>(color.r)));
+        params.push_back(boost::lexical_cast<std::string>(static_cast<int>(color.g)));
+        params.push_back(boost::lexical_cast<std::string>(static_cast<int>(color.b)));
+        params.push_back(boost::lexical_cast<std::string>(static_cast<int>(color.a)));
+
+        AddOpenTag("rgba", &params);
+    }
+
+private:
+    const Font& m_font;
+    std::string m_text;
+    std::vector<boost::shared_ptr<TextElement> > m_text_elements;
+    bool m_are_widths_calculated;
+};
+
+
+Font::TextElements::TextElements(const Font& font) :
+    pimpl(new TextElementsImpl(font))
+{}
+// Required because Impl is defined here
```

Comment:
add blank lines between } of one function and class name of the next (or comment)